fix: correct sorted listings with product variants and different prices

ulloe opened the pull request:
Fixes an issue where product variants with different prices were not correctly included in the price sorting.

targets #4053

<!--
Thank you for contributing to Shopware! Please fill out this description template to help us process your pull request.

Please make sure to fulfil our contribution guidelines (https://developer.shopware.com/docs/resources/guidelines/code/contribution?category=shopware-platform-dev-en/contribution).

Do your changes need to be documented?
Please create a second pull request at https://github.com/shopware/docs
-->

### 1. Why is this change necessary?
See #4053 

If we have Product P1 priced at 10 EUR, and Product P2, which has variants priced at 5 EUR and 15 EUR, and we sort the listings by price (asc), P2 (displayed as "From 5 EUR") appears after P1 (10 EUR). This is not the expected behavior.

### 2. What does this change do, exactly?
It adds a MIN() function to the accessor when sorting by cheapestPrice is selected.

### 3. Describe each step to reproduce the issue or behaviour.
1. Create a product with any price.
2. Create another product.
3. Add two variants to the second product: one variant priced lower than the first product, and the other priced higher.
4. Visit a page with a listing that includes both products. Sort the listing by price in ascending order. The second product appears after the first one, even though its displayed price is lower.

### 4. Please link to the relevant issues (if any).
- closes #4053 

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [X] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [X] I have written or adjusted the documentation according to my changes
- [X] This change has comments for package types, values, functions, and non-obvious lines of code
- [X] I have read the contribution requirements and fulfilled them

Commit:
fix: correct sorted listings with product variants and different prices

Fixes an issue where product variants with different prices were not correctly included in the price sorting.

targets #4053

mitelg committed:
Merge branch 'trunk' into variant-price-sorting-criteria

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/shopware/shopware?pullRequest=11185) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/shopware/shopware?pullRequest=11185) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: ulloe<br/>:white_check_mark: mitelg<br/>:x: Ullrich Löblein<br/><hr/>**Ullrich Löblein** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/shopware/shopware?pullRequest=11185) it.</sub>

mitelg commented on src/Core/Framework/DataAbstractionLayer/Dbal/CriteriaQueryBuilder.php:
```diff
@@ -146,6 +146,8 @@ public function addSortings(EntityDefinition $definition, Criteria $criteria, ar
                     $accessor = 'MAX(' . $accessor . ')';
                 }
+            } else {
+                $accessor = 'MIN(' . $accessor . ')';
```
maybe we can add this before going into the if `if (!\in_array($sorting->getField(), ['product.cheapestPrice', 'cheapestPrice'], true)) {` and then simply check for `FieldSorting::DESCENDING` where we add the `MAX()` function. 


also a test would be nice.

ulloe replied:
Please disregard this if I have misunderstood, but keep in mind: the accessor should always be `MIN()` for the `cheapestPrice` option. Even if I sort by the cheapest price descending, the cheapest price still serves as the reference.

mitelg replied:
```
$accessor = 'MIN(' . $accessor . ')';
if (!\in_array($sorting->getField(), ['product.cheapestPrice', 'cheapestPrice'], true)) {
    if ($sorting->getDirection() === FieldSorting::DESCENDING) {
        $accessor = 'MAX(' . $accessor . ')';
    }
}
```

this is what I meant :blush: does it make sense for you?

ulloe replied:
No it does not, I'm sorry. 🙃 
Keep in mind, that we concatenate the `$accessor` . Your suggestion can cause `SQLSTATE[HY000]: General error: 1111 Invalid use of group function` errors. There are cases in which we end up with an `MAX(MIN( $accessor ))` accessor.

I'd stick to the `if / else` solution or maybe use something like:

```php
if (\in_array($sorting->getField(), ['product.cheapestPrice', 'cheapestPrice'], true)) {
    $accessor = 'MIN(' . $accessor . ')';
} else if ($sorting->getDirection() === FieldSorting::DESCENDING) {
    $accessor = 'MAX(' . $accessor . ')';
} else {
    $accessor = 'MIN(' . $accessor . ')';
}
```

mitelg replied:
oh right :man_facepalming: would work with renaming the variable :grin: but okay, your solution works

mitelg commented on changelog/_unreleased/2025-07-11-fix-correctly-sort-product-variants-with-different-prices.md:
```diff
@@ -0,0 +1,15 @@
+---
+title: Adds MIN() Function to cheapestPrice Accessor
+issue: #4053
+flag:
```
```suggestion
```

ulloe committed:
Update changelog/_unreleased/2025-07-11-fix-correctly-sort-product-variants-with-different-prices.md

Co-authored-by: Michael Telgmann <mitelg@users.noreply.github.com>

mitelg commented on src/Core/Framework/DataAbstractionLayer/Dbal/CriteriaQueryBuilder.php:
```diff
@@ -146,6 +146,8 @@ public function addSortings(EntityDefinition $definition, Criteria $criteria, ar
                     $accessor = 'MAX(' . $accessor . ')';
                 }
+            } else {
+                $accessor = 'MIN(' . $accessor . ')';
```
but we still would need a test for this

ulloe replied:
I have created and committed a test. However, please keep in mind that the test _may_ also pass without the fix. As described in Issue #4053 and the [MySQL documentation](https://dev.mysql.com/doc/refman/8.4/en/group-by-handling.html#:~:text=unless%20they%20are%20the%20same%2C%20the%20values%20chosen%20are%20nondeterministic%2C%20which%20is%20probably%20not%20what%20you%20want.). The result of the final SQL query is not deterministic in this case.

Commit:
test: sorted listings with product variants and different prices
This test includes a set of valid and invalid IDs.
As described in the MySQL documentation (https://dev.mysql.com/doc/refman/8.4/en/group-by-handling.html), when ONLY_FULL_GROUP_BY is disabled, "the values chosen are nondeterministic." This behavior makes it difficult to create a reliable test that consistently fails without the corresponding fix.
As a result, this test might pass even without using the MIN() function in the accessor.

mitelg commented on tests/integration/Core/Framework/DataAbstractionLayer/Dbal/CriteriaQueryBuilderTest.php:
```diff
@@ -0,0 +1,98 @@
+        'p1.2' => '0198bd446e077084984f95175b2bea27',
+    ];
+
+    private const INVALID_TEST_SET = [
```
this is unused.

ulloe replied:
I wanted to keep a set of IDs that does not produce the expected behavior due to MySQL's non-deterministic results when `ONLY_FULL_GROUP_BY` is disabled. I've now moved that ID set into the comment for clarity, since it's not used in the actual test. 
I'd like to add to this comment, if that's okay with you.

```php
/**
 * This test checks listing sorting behavior affected by MySQL's GROUP BY handling.
 *
 * Shopware disables ONLY_FULL_GROUP_BY, allowing queries that may return non-deterministic
 * results when selecting columns not functionally dependent on the GROUP BY clause.
 * See: https://dev.mysql.com/doc/refman/8.4/en/group-by-handling.html
 *
 * The CriteriaQueryBuilder generates such a query. Without the fix, results may vary between
 * runs; with the fix, the outcome is deterministic and the test should always pass.
 *
 * A specific ID set is used to reproduce the issue. For reference, the following ID set
 * does NOT trigger the problem and is noted here for completeness:
 * ['s1' => '00000000000000000000000000000001',
 *  'p1' => '00000000000000000000000000000002',
 *  'p1.1' => '00000000000000000000000000000003',
 *  'p1.2' => '00000000000000000000000000000004']
 *
 * Changing the data may prevent the issue from appearing, so edit with caution.
 */
```

mitelg commented on tests/integration/Core/Framework/DataAbstractionLayer/Dbal/CriteriaQueryBuilderTest.php:
```diff
@@ -0,0 +1,98 @@
+class CriteriaQueryBuilderTest extends TestCase
+{
+    use IntegrationTestBehaviour;
+    private const VALID_TEST_SET = [
```
why putting them into a private constant at all?

ulloe replied:
From my perspective, there's no real need to keep these IDs in a private constant. I'm happy to remove it. I'll include everything in the next commit.

mitelg commented on tests/integration/Core/Framework/DataAbstractionLayer/Dbal/CriteriaQueryBuilderTest.php:
```diff
@@ -0,0 +1,98 @@
+
+    public IdsCollection $ids;
+
+    public function testSortingByCheapestPrice(): void
```
wouldn't it be possible to recreate the situation described in the issue?

ulloe replied:
I'm not entirely sure what’s being asked here. I _did_ recreate the situation described in the original issue - that’s what this test is based on.
If the question is about reproducing the issue with two articles and their variants, then yes, I can easily add that setup.
However, if the concern is about making the test consistently fail without the fix, that’s unfortunately not possible. The non-deterministic nature of the results (due to MySQL’s internal behavior and the lack of `ONLY_FULL_GROUP_BY`) makes that inherently unreliable without the fix in place.

mitelg replied:
okay, I understand :+1:

Commit:
test: Update CriteriaQueryBuilderTest.php

* Add test to compare two products with variants (related to issue #4053)
* Add test case for descending order direction
* Add comment for completeness

mitelg commented:
thanks for your contribution @ulloe 👍🎉💙
